### PR TITLE
Support zero-size vertex and index buffer bindings

### LINKED ISSUE #3170: Support zero-length vertex/index buffers
The WebGPU specification says that [`setVertexBuffer` and `setIndexBuffer` may be passed a size of zero](https://www.w3.org/TR/webgpu/#dom-gpurendercommandsmixin-setvertexbuffer), but the wgpu Rust API doesn't support this: the `set_index_buffer` and `set_vertex_buffer` methods take `BufferSlice` values, which are not allowed to be empty.

The wgpu-core APIs (used by Firefox and Deno) allow zero-size vertex/index buffers and bindings.

As explained in the documentation for `hal::BufferBinding`, some backends do not support zero-size bindings. Previously, this documentation stated that a zero-size `hal::BufferBinding` was not permitted, but this restriction was not honored elsewhere in the code. To handle zero-size bindings propertly, they need to be quashed or remapped to a non-zero size, either universally in wgpu-core, or in the specific backends that do not support them.

An incomplete attempt to clean up the handling of zero-length buffers was made in #7875 (later reverted in #7905) and #7911. The relevant code (in particular, the `ShouldBeNonZeroExt` trait and the type of the `size` member of `hal::BufferBinding`) should be cleaned up when this issue is addressed.

Note that although WebGPU allows zero-size vertex and index buffers, [it does not allow zero-size buffer bindings in a bind group](https://www.w3.org/TR/webgpu/#dom-gpudevice-createbindgroup):

> The bound part designated by bufferBinding.[offset](https://www.w3.org/TR/webgpu/#dom-gpubufferbinding-offset) and bufferBinding.[size](https://www.w3.org/TR/webgpu/#dom-gpubufferbinding-size) resides inside the buffer and has non-zero size.

### PR DESCRIPTION
**Pull request by andyleiserson**
This started as the follow-up to #9791 to fix cases where `hal` resolved bindings using the allocated size of a buffer instead of the application-specified size. Some things clicked for me that I hadn't fully made sense of when I originally looked at #8034, #7875, and related items, and I decided it made sense to also fix #3170 as part of this cleanup.

This is structured as three commits:
* The first commit parameterizes the type of `hal::BufferBinding::size` (to be `BufferSize` (`NonZeroU64`) for storage/uniform bindings, and `BufferAddress` (`u64`) for vertex/index bindings), and resolves binding sizes in core instead of hal. This commit temporarily breaks some zero-size vertex/index binding cases that worked before, the third commit fixes them.
* The second commit adds directed tests for zero-size bindings. The subtlety these tests are intended to catch is that `vkCmdBindVertexBuffers` and `vkCmdBindIndexBuffer` do not specify the length of a binding, so an implementation that passed the offset of a zero-size binding through to the Vulkan hal as-is, could incorrectly expose non-zero data to shaders. So a zero-size binding anywhere in the buffer, needs to be relocated to the end of the buffer, before passing it to the Vulkan hal. (We do it unconditionally, since the other hals don't care.)
* The third commit fixes zero-size vertex and index buffer bindings. This is mostly a matter of changing `BufferSize` to `BufferAddress` in a bunch of places.

**Testing**
Adds directed tests. This change also resolves a vulkan validation error in the CTS test `webgpu:api,validation,encoding,cmds,render,setIndexBuffer:*`, but we don't fail on vulkan validation errors when we run the CTS in CI (although we probably should; related: #8094), so this doesn't show up as an expectation change.

**Squash or Rebase?** Rebase (after squashing fixups)

**Checklist**

- [x] I self-reviewed and fully understand this PR.
- [x] WebGPU implementations built with `wgpu` may be affected behaviorally.
- [ ] Validation and feature gates are in place to confine behavioral changes.
- [x] Tests demonstrate the validation and altered logic works. 
- [x] `CHANGELOG.md` entries for the user-facing effects of this change are present. 
- [ ] The PR is minimal, and doesn't make sense to land as multiple PRs.
  - It could be split in two, but that would require temporarily breaking things on trunk, or handling the intermediate state more carefully.
- [x] Commits are logically scoped and individually reviewable.
- [x] The PR description has enough context to understand the motivation and solution implemented.

### CODE REVIEW

File: wgpu-hal/src/metal/device.rs [hunk 1]
```diff
@@ -449,9 +449,9 @@ impl super::Device {
 
     pub unsafe fn buffer_from_raw(
         raw: Retained<ProtocolObject<dyn MTLBuffer>>,
-        size: wgt::BufferAddress,
+        _size: wgt::BufferAddress,
```
Comment (andyleiserson) [inline wgpu-hal/src/metal/device.rs]: Not sure if it makes sense to remove this, or keep it in case there is a reason for `hal` to know about buffer sizes in the future (that seems a little dangerous, since passing incorrect sizes will silently work in the interim, although they would have silently worked in most cases in the past as well).

File: wgpu/src/dispatch.rs [hunk 2]
```diff
@@ -546,14 +546,14 @@ pub trait RenderBundleEncoderInterface: CommonTraits {
...
-        size: Option<crate::BufferSize>,
+        size: Option<crate::BufferAddress>,
     );
     fn set_vertex_buffer(
         &mut self,
         slot: u32,
         buffer: Option<&DispatchBuffer>,
         offset: crate::BufferAddress,
-        size: Option<crate::BufferSize>,
+        size: Option<crate::BufferAddress>,
```
Comment (sagudev) [inline wgpu/src/dispatch.rs:587]: It is awkward that size has type of address. After this PR lands it might make sense to just inline the type aliases for size as that would be more clear IMO.

Comment (andyleiserson) [reply wgpu/src/dispatch.rs:587]: I think there is value in having a type alias for "integer that can address WebGPU buffers", but maybe we could have `BufferScaleInteger` and `NonZero<BufferScaleInteger>` rather than trying to capture whether it can be zero in the name. (I'm not actually proposing the name `BufferScaleInteger`, but I also don't have something better to offer. I don't think `BufferAddress` is terrible as long as we don't use it for sizes while also having `BufferSize`.)

Comment (sagudev) [reply wgpu/src/dispatch.rs:587]: `BufferSize` and `NonZero<BufferSize>` 🙃

File: tests/tests/wgpu-gpu/zero_init.rs [hunk 3]
```diff
@@ -1563,3 +1567,471 @@ async fn check_vertex_buffer_tail_init(
...
+    // Choose which buffer is bound zero-size, and how many indices to draw. The
+    // index-buffer case can only issue a zero-index draw, because a zero-size
+    // index binding yields an index limit of zero.
+    let (vb_slice, ib_slice, index_count) = match case {
+        ZeroBindingCase::VertexAtEnd => (
+            vertex_buffer.slice(vb_size..vb_size),
+            index_buffer.slice(..),
+            ZB_VERT_COUNT,
+        ),
+        ZeroBindingCase::VertexNotAtEnd => (
```
Comment (Wumpf) [inline tests/tests/wgpu-gpu/zero_init.rs]: Why not call this `VertexAtBeginning` since we're starting at vertex zero and then do out of bounds read from there?

File: tests/tests/wgpu-gpu/zero_init.rs [hunk 4]
```diff
@@ -1563,3 +1567,471 @@ async fn check_vertex_buffer_tail_init(
...
+            vertex_buffer.slice(0..0),
+            index_buffer.slice(..),
+            ZB_VERT_COUNT,
+        ),
+        ZeroBindingCase::IndexAtEnd => (
```
Comment (Wumpf) [inline tests/tests/wgpu-gpu/zero_init.rs:3348]: since we end up specifying zero indices, we're not actually performing any vertex or indexb buffer read at all, no? 🤔 Ofc still relevant to test zero indices, but that's technically not a zero init test then, is it? I.e. this is more of a `ZeroBindingCase::ZeroIndices`?

File: wgpu-core/src/device/resource.rs [hunk 5]
```diff
@@ -1135,22 +1139,32 @@ impl Device {
...
+        //    allow.
+        //  - The overall buffer size must generally be a non-zero multiple of
+        //    COPY_BUFFER_ALIGNMENT (4), but in special cases an additional platform-dependent
+        //    requirement may apply (e.g. 256 for D3D12 constant buffers).
+        //
+        // Because initialization operates at multiples of COPY_BUFFER_ALIGNMENT, and
+        // initialization tracking will never determine that it is necessary to initialize a
+        // region outside the application-visible range, we adjust the application-requested
+        // size downwards to a multiple of COPY_BUFFER_ALIGNMENT, and eagerly zero-initialize
+        // anything beyond that.
```
Comment (Wumpf) [inline wgpu-core/src/device/resource.rs]: nit: found it a bit confusing since the application requested size isn't what we adjust downwards, it's the `tail_start` rather
```suggestion
        // region outside the application-visible range, round the tracked initialization range
        // down to a multiple of COPY_BUFFER_ALIGNMENT (tail_start),  and eagerly zero-initialize
        // everything from there on.
```

File: wgpu-core/src/resource.rs [hunk 6]
```diff
@@ -597,36 +597,68 @@ impl Buffer {
...
+
+    /// Resolve the binding range for a vertex or index buffer.
+    ///
+    /// These may have zero size. For storage and uniform buffer bindings, which
+    /// must have non-zero size, use [`Buffer::binding`].
+    ///
+    /// Returns an error if the binding would overflow the buffer.
+    ///
+    /// Vulkan's `vkCmdBindVertexBuffers` and `vkCmdBindIndexBuffer` have
+    /// some important restrictions:
```
Comment (Wumpf) [inline wgpu-core/src/resource.rs]: bit uneasy about calling out Vulkan in core, would be nice if we could just declare this as the contract of hal rather and motivate it there

File: wgpu-hal/src/metal/command.rs [hunk 7]
```diff
@@ -1421,13 +1421,14 @@ impl crate::CommandEncoder for super::CommandEncoder {
...
-                core::num::NonZeroU64::new(buffer_size).unwrap(),
-            );
+        // The size map uses `NonZeroU64` mostly because it did so in the past when
+        // `hal::BufferBinding` tried to disallow zero-length bindings. Zero-length bindings
+        // are valid for vertex buffers, so there isn't really a significance of a
+        // zero-length buffer being removed from the map (vs. the map storing `u64`).
+        if let Some(size) = NonZeroU64::new(binding.size) {
+            self.state.vertex_buffer_size_map.insert(buffer_index, size);
         } else {
+            // `make_sizes_buffer_update` uses zero for buffers that are not in the map.
```
Comment (Wumpf) [inline wgpu-hal/src/metal/command.rs]: what's the reason not to refactor this to just store `wgt::BufferAddress`? Sure, as explained it works either way, but as pointed out it's legacy complexity. Left out because too much churn?

File: wgpu-hal/src/metal/device.rs (hunk 1 above)
Comment (Wumpf) [reply wgpu-hal/src/metal/device.rs]: I'm very much in favor of removing sizes entirely from these. This would cleanly then reflects that at no point in time we validate these sizes directly as that's now the new (~ weakened) contract of hal.
It's quite interesting that we no longer track buffer sizes anywhere in hal now, but it makes sense!

File: wgpu-core/src/command/render.rs [hunk 8]
```diff
@@ -2897,24 +2900,27 @@ fn set_index_buffer(
...
+            range.clone(),
             MemoryInitKind::NeedsInitializedMemory,
         ),
     );
 
     unsafe {
         hal::DynCommandEncoder::set_index_buffer(
             state.pass.base.raw_encoder,
-            binding,
+            hal::BufferBinding::new_unchecked(raw, range.start, range.end),
```
Comment (Wumpf) [inline wgpu-core/src/command/render.rs]: I think this is a bug - the third parameter of `BufferBinding` is size, not end:

```suggestion
            hal::BufferBinding::new_unchecked(raw, range.start, range.end - range.start),
```
(I don't think it needs to be a checked sub since `resolve_vertex_or_index_binding_range` should have already made sure that it's fine)

File: wgpu-hal/src/dx12/device.rs [hunk 9]
```diff
@@ -503,26 +503,33 @@ impl super::Device {
...
     ) -> Result<super::Buffer, crate::DeviceError> {
-        let mut desc = desc.clone();
-
         if desc.usage.contains(wgt::BufferUses::UNIFORM) {
-            desc.size = desc
+            debug_assert!(desc
                 .size
-                .next_multiple_of(Direct3D12::D3D12_CONSTANT_BUFFER_DATA_PLACEMENT_ALIGNMENT.into())
+                .is_multiple_of(Direct3D12::D3D12_CONSTANT_BUFFER_DATA_PLACEMENT_ALIGNMENT.into()))
         }
```
Comment (Wumpf) [inline wgpu-hal/src/dx12/device.rs:516]: my llm pointed out that `ray-traced-triangle` & `halmark` actually rely on this right now and will fail this debug assertion. Confirmed via
* `cargo run -p wgpu-hal --example ray-traced-triangle --features dx12`
* `cargo run -p wgpu-hal --example halmark --features dx121`

(halmark aligns one of its uniform buffers but not `Globals`)

We should be nice and patch that up

File: wgpu-core/src/resource.rs [hunk 10]
```diff
@@ -597,36 +597,68 @@ impl Buffer {
...
+        &self,
+        offset: wgt::BufferAddress,
+        size: Option<wgt::BufferAddress>,
+    ) -> Result<Range<wgt::BufferAddress>, BindingError> {
+        let resolved_size = self.resolve_binding_size(offset, size)?;
+        if resolved_size != 0 {
+            Ok(offset..offset + resolved_size)
+        } else {
+            // Relocate zero-size binding to end of buffer. See above.
+            Ok(self.size..self.size)
```
Comment (Wumpf) [inline wgpu-core/src/resource.rs]: llm found an interesting corner case I didn't think of: index buffer bindings must be index aligned, and this may not actually be: there doesn't seem to be anything that prevents a 6byte buffer to be your uint32 index buffer 😱 
Therefore this would land 6..6 which then violates [`VUID-vkCmdBindIndexBuffer-offset-00432`](https://docs.vulkan.org/refpages/latest/refpages/source/vkCmdBindIndexBuffer.html#VUID-vkCmdBindIndexBuffer-offset-08783).
Should probably add a test for using an index buffer with an unaligned size

**Review by Wumpf — CHANGES_REQUESTED** [review]
impressive test suite, quite intricate. It's nice to see how strong the protections are even with indirect draw validation disabled (even though that isn't the point of the test).

Found some issues with llm help (all of which I pruned and validated to my knowledge ofc :)) which do look blocking to me.

File: wgpu-core/src/command/render.rs (hunk 8 above)
Comment (andyleiserson) [reply wgpu-core/src/command/render.rs]: Good catch, thank you!

File: tests/tests/wgpu-gpu/zero_init.rs (hunk 4 above)
Comment (andyleiserson) [reply tests/tests/wgpu-gpu/zero_init.rs:3348]: On Vulkan we do actually read bytes from the buffer, because zero-size bindings aren't supported. So we give a 4-byte binding of zeros, from which reads have the same observable result as a truly out-of-bounds read (and, prior to #9791, were the subject of an actual zero-init bug).

I do see how these tests don't totally fit in this file though. Thoughts on a better place to put them? Given the clarification, do you still think `ZeroIndices` is better than `IndexAtEnd`.

I'll also update the comment before these tests (~line 1575) to say explicitly that this is what we're doing with zero-size bindings.

File: wgpu-hal/src/metal/command.rs (hunk 7 above)
Comment (andyleiserson) [reply wgpu-hal/src/metal/command.rs]: I was just trying to avoid scope creep, but the change is really small and it makes things less confusing, so I'll add it to the PR.

File: wgpu-hal/src/dx12/device.rs (hunk 9 above)
Comment (andyleiserson) [reply wgpu-hal/src/dx12/device.rs:516]: Rather than have to add a call to `buffer_size_alignment` before every `create_buffer`, I changed `create_buffer` to return the actual allocated size. I attempted to also move the vertex/index buffer padding case into hal, but immediately found that Metal also requires `offset < size`. So to change fewer things and reduce the risk of breakage, I moved the vertex/index buffer padding case back into core (which is a bit contrary to the goal of not sprinkling details about platform special cases throughout core, but was also already the case before this PR).

File: wgpu-core/src/resource.rs (hunk 10 above)
Comment (andyleiserson) [reply wgpu-core/src/resource.rs]: Another good catch, fixed. This also required changing the buffer size calculation again, now it is:

```rust
// We require 4B of padding with 4B alignment. See above.
align_to(desc.size - 1, wgt::COPY_BUFFER_ALIGNMENT) + wgt::COPY_BUFFER_ALIGNMENT
```

Instead of `align_to(desc.size + 1, wgt::COPY_BUFFER_ALIGNMENT)`.

**Review by Wumpf — CHANGES_REQUESTED** [review]
looking great, thanks for going deep on all the comments and apologies for the long delay!

Unfortunately I was able to pin point another bug cornercase with some llm help:
* allocating a vertex buffer of size 5 gives us 8 bytes buffer size ( `align_to( 5 - 1, 4) + 4 == 8`)
* `pass.set_vertex_buffer(0, buffer.slice(0..0));` pushes the zero size binding at `align_to(5, 4) == 8` (in `resolve_vertex_or_index_binding_range`)
-> end up reading at the end

This failure can easily be proven by changing from 2 extra bytes to 1 extra byte in `create_zero_binding_buffer`. Doing so I get test failures: `with code 0xc0000409: The system detected an overrun of a stack-based buffer in this application. This overrun could potentially allow a malicious user to gain control of this application. (os error 1282)`

Probably best to compute where `resolve_vertex_or_index_binding_range` would end up and then add a `wgt::COPY_BUFFER_ALIGNMENT` there

**Comment by andyleiserson** [comment]
I... am not sure what I was thinking there. Fixed it, and extended the tests to cover all the mod-4 alignments. I also did my own LLM review pass which identified a bunch of stuff.

There's four new commits, and generally a lot of fixups that I would definitely like to squash before merging.

File: wgpu-core/src/resource.rs [hunk 11]
```diff
@@ -622,10 +622,21 @@ impl Buffer {
             None => buffer_size
                 .checked_sub(offset)
                 .and_then(|remaining| S::try_from(remaining).ok())
-                .ok_or_else(|| BindingError::BindingOffsetTooLarge {
-                    buffer: self.error_ident(),
-                    offset,
-                    buffer_size,
+                .ok_or_else(|| {
+                    if offset <= buffer_size {
+                        debug_assert!(offset == buffer_size);
```
Comment (Wumpf) [inline wgpu-core/src/resource.rs]: nit: that's a thing, right?
```suggestion
                        debug_assert_eq!(offset, buffer_size);
```

File: wgpu-core/src/command/render.rs [hunk 12]
```diff
@@ -2854,11 +2854,17 @@ fn set_vertex_buffer(
...
             .into());
         }
-        if let Some(size) = size {
-            return Err(RenderCommandError::from(
-                crate::binding_model::BindingError::UnbindingVertexBufferSizeNotZero { slot, size },
-            )
-            .into());
+        match size {
+            Some(size) if size != 0 => {
+                return Err(RenderCommandError::from(
```
Comment (Wumpf) [inline wgpu-core/src/command/render.rs:2932]: nit: why not `if let Some(size) = size && size != 0`

**Review by Wumpf — APPROVED** [review]
thanks should be all good now 🤞

File: wgpu-core/src/command/render.rs (hunk 12 above)
Comment (andyleiserson) [reply wgpu-core/src/command/render.rs:2932]: It requires edition 2024. Claude also suggested `size.is_some_and(|size| != 0)`, but the error variant wants the non-`Option` size.

**Comment by andyleiserson** [comment]
This hasn't moved yet because I had been waiting to clear the Firefox `wgpu` update before merging it, and that has taken longer than expected.

File: wgpu-core/src/command/render.rs (hunk 12 above)
Comment (Wumpf) [reply wgpu-core/src/command/render.rs:2932]: ah didn't realize we still don't have edition 2024
